File: leetcode/994_rotting_oranges.py
from typing import List, Tuple


class Solution:

    EMPTY, FRESH, ROTTEN = range(3)

    def orangesRotting(self, grid: List[List[int]]) -> int:
        if not any(grid):
            return 0
        result = 0
        fresh_count = 0
        rotten = []
        for y, row in enumerate(grid):
            for x, val in enumerate(row):
                if val == self.FRESH:
                    fresh_count += 1
                elif val == self.ROTTEN:
                    rotten.append((y, x))
        if not fresh_count:
            return result
        time = 0
        while rotten:
            if not fresh_count:
                break
            time += 1
            new_rotten = []  # type: List[Tuple[int, int]]
            for y, x in rotten:
                fresh_near = self.get_fresh_near(grid, y, x)
                for yy, xx in fresh_near:
                    fresh_count -= 1
                    grid[yy][xx] = self.ROTTEN
                new_rotten.extend(fresh_near)
            rotten = new_rotten
        return time if not fresh_count else -1

    def get_fresh_near(self, grid, y, x):
        fresh = []
        rows = len(grid)
        cols = len(grid[0])
        for dy, dx in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
            near_y, near_x = y + dy, x + dx
            if 0 <= near_y < rows and 0 <= near_x < cols:
                if grid[near_y][near_x] == self.FRESH:
                    fresh.append((near_y, near_x))
        return fresh

File: leetcode/java/994RottingOranges.java
/**
 * T: O(R*C) S: O(R*C)
 *
 * <p>Find all rotten oranges, iterate over all of their neighbors that will rot and remember their
 * rot time. If after that there are still fresh oranges - return -1; otherwise return maximum seen
 * rot time(if any).
 */
import java.util.ArrayDeque;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Queue;

class Solution {
  int EMPTY_CELL = 0;
  int FRESH_ORANGE = 1;
  int ROTTEN_ORANGE = 2;
  int[] dx = new int[] {-1, 0, 1, 0};
  int[] dy = new int[] {0, -1, 0, 1};

  public int orangesRotting(int[][] grid) {
    int rows = grid.length;
    int cols = grid[0].length;
    Queue<Integer> rottenOranges = new ArrayDeque<>() {};
    Map<Integer, Integer> depth = new HashMap<>();
    for (int y = 0; y < rows; y++) {
      for (int x = 0; x < cols; x++) {
        if (grid[y][x] == ROTTEN_ORANGE) {
          int cellCode = y * cols + x;
          rottenOranges.offer(cellCode);
          depth.put(cellCode, 0);
        }
      }
    }
    while (!rottenOranges.isEmpty()) {
      int currentCellCode = rottenOranges.poll();
      int currentY = currentCellCode / cols;
      int currentX = currentCellCode % cols;
      for (int direction = 0; direction < dx.length; direction++) {
        int nextY = currentY + dy[direction];
        int nextX = currentX + dx[direction];
        if (nextY >= 0
            && nextY < rows
            && nextX >= 0
            && nextX < cols
            && grid[nextY][nextX] == FRESH_ORANGE) {
          grid[nextY][nextX] = ROTTEN_ORANGE;
          int nextCellCode = nextY * cols + nextX;
          rottenOranges.offer(nextCellCode);
          depth.put(nextCellCode, depth.get(currentCellCode) + 1);
        }
      }
    }
    for (int[] row : grid) {
      for (int cell : row) {
        if (cell == FRESH_ORANGE) {
          return -1;
        }
      }
    }
    return depth.size() == 0 ? 0 : Collections.max(depth.values());
  }
}
